cmd/kubeadm/app/cmd/join.go

```diff
@@ -235,34 +279,228 @@ func (j *Join) Validate(cmd *cobra.Command) error {
+		// the deployment of control plane components on the node.
+		err := j.MarkMaster()
+		if err != nil {
+			return err
+		}
+
+		// outputs the join master done template and exits
+		ctx := map[string]string{
+			"KubeConfigPath": kubeadmconstants.GetAdminKubeConfigPath(),
+		}
+		joinMasterNodeDoneTemp.Execute(out, ctx)
+		return nil
+	}
+
+	// otherwise, if the node joined as a worker node
+
+	// outputs the join done message and exits
+	fmt.Fprintf(out, joinWorkerNodeDoneMsgf)
+	return nil
+}
+
+// FetchInitClusterConfiguration reads the master configuration from the kubeadm-admin configMap,
+// or, as fallback strategy, it loads the master configuration from the given MasterConfigurationFile path.
+// Nb. The fallback strategy addresses scenarios where authorizations for bootstrap tokens
+// will be explicitly revoked after kubeadm init
+func (j *Join) FetchInitClusterConfiguration(tlsBootstrapCfg *clientcmdapi.Config) (*kubeadmapi.MasterConfiguration, error) {
+
+	// creates a client to access the cluster using the bootstrap token identity
+	tlsClient, err := kubeconfigutil.ToClientSet(tlsBootstrapCfg)
+	if err != nil {
+		return nil, fmt.Errorf("error creating the kubernetes client with tls bootstrap authentity: %v", err)
+	}
+
+	// Fetches the master configuration
+	// ** Note for the reviewers:
+	// ** see notes in phases/upgrade about how to share FetchConfiguration between phase/update and join
+	masterConfigurationExt, err := upgrade.FetchConfiguration(tlsClient, os.Stdout, j.cfg.MasterConfigurationFile)
+	if err != nil {
+		return nil, fmt.Errorf("error reading the master configuration for the cluster: %v", err)
+	}
+
+	// Converts public API to internal API
+	masterConfiguration := &kubeadmapi.MasterConfiguration{}
+	legacyscheme.Scheme.Convert(masterConfigurationExt, masterConfiguration, nil)
 
-	// Write the bootstrap kubelet config file or the TLS-Boostrapped kubelet config file down to disk
-	if err := kubeconfigutil.WriteToDisk(kubeconfigFile, cfg); err != nil {
+	return masterConfiguration, nil
+}
+
+// CheckIfReadyForJoinAsMaster ensures that the cluster is in a configuration that supports
+// joining additional masters
+func (j *Join) CheckIfReadyForJoinAsMaster(masterConfiguration *kubeadmapi.MasterConfiguration) error {
+
+	// checks if the cluster was created with features.HighAvailability enabled
+	if !features.Enabled(masterConfiguration.FeatureGates, features.HighAvailability) {
+		return fmt.Errorf("unable to join a new master node on a cluster that wasn't created without '--feature-gates=HighAvailability=true'")
+	}
+
+	// ** Note for reviewers:
+	// ** If the cluster was initialized with --feature-gates=HighAvailability=true following checks were already executed
+	// ** during kubeadm init:
+	// ** 1. the cluster uses an external etcd
+	// ** 2. the cluster uses an external load balancer
+	// ** Should we double check the above assumptions before joining?
```

Comment:
TBD.

Reply:
IMO the key point here is that it is required to expand the default grants that aree assigned to bootstrap tokens to allow to check the real cluster conditions in this point, and this could raise some security concerns.

Since I prefer to avoid any security concern, I think we should accept the consequent trade-off, that is the fact that when joining a new master we assume that the cluster is still in the conditions we ensured while doing `kubeadm init --feature-gates=HighAvailability`

WDYT?